hypothesis-python/src/hypothesis/internal/conjecture/engine.py

```diff
@@ -992,12 +1003,34 @@ def add(self, data):
             self.best_status = data.status
             self.reset()
 
-        self.fresh_examples.append(data)
-        if len(self) > self.pool_size:
-            pop_random(self.random, self.used_examples or self.fresh_examples)
-            assert self.pool_size == len(self)
+        for label in data.target_observations:
+            pool = self.scored_examples[label]
+            pool.append(data)
+            pool.sort(key=lambda d: d.target_observations[label], reverse=True)
+
+        if self.scored_examples:
+            self.fresh_examples = []
+            self.used_examples = []
+            for pool in self.scored_examples.values():
+                # We have to check each pool, not just those we added `data` to,
+                # because we may have just found a new label and started a new pool.
```

Comment:
I don't understand this comment. If we'd done that wouldn't that be a pool we'd added data to?

Reply:
I've rewritten the comment... Because the size cap is divided between known labels, finding a new label decreases the per-pool size cap.  Existing pools might now be too large, so we recheck everything.

Tracking a lot more state *could* make this faster (or slower!), but definitely more error-prone.  I've tried to go with "obviously no errors" for this round, since I expect that we'll replace the implementation a few times anyway.